src/frequenz/client/dispatch/_client.py

```diff
@@ -91,45 +95,61 @@ async def list(
             end_to: optional, filter by end_time < end_to.
             active: optional, filter by active status.
             dry_run: optional, filter by dry_run status.
+            page_size: optional, number of dispatches to return per page.
 
         Returns:
-            An async iterator of dispatches.
-        """
-        start_time_interval = None
-        end_time_interval = None
-
-        if start_from or start_to:
-            start_time_interval = PBTimeIntervalFilter(
-                **{"from": to_timestamp(start_from)}, to=to_timestamp(start_to)
-            )
+            An async iterator over pages of dispatches.
 
-        if end_from or end_to:
-            end_time_interval = PBTimeIntervalFilter(
-                **{"from": to_timestamp(end_from)}, to=to_timestamp(end_to)
-            )
+        Yields:
+            A page of dispatches.
 
-        selectors = []
+        """
 
-        for selector in component_selectors:
-            selectors.append(component_selector_to_protobuf(selector))
+        def to_interval(
+            from_: datetime | None, to: datetime | None
+        ) -> PBTimeIntervalFilter | None:
+            return (
+                PBTimeIntervalFilter(
+                    **{"from": to_timestamp(from_)}, to=to_timestamp(to)
+                )
+                if from_ or to
+                else None
+            )
 
+        # Setup parameters
+        start_time_interval = to_interval(start_from, start_to)
+        end_time_interval = to_interval(end_from, end_to)
+        selectors = list(map(component_selector_to_protobuf, component_selectors))
         filters = DispatchFilter(
             selectors=selectors,
             start_time_interval=start_time_interval,
             end_time_interval=end_time_interval,
             is_active=active,
             is_dry_run=dry_run,
         )
+
         request = ListMicrogridDispatchesRequest(
-            microgrid_id=microgrid_id, filter=filters
+            microgrid_id=microgrid_id,
+            filter=filters,
+            pagination_params=PaginationParams(page_size=page_size),
         )
 
-        response = await cast(
-            Awaitable[ListMicrogridDispatchesResponse],
-            self._stub.ListMicrogridDispatches(request, metadata=self._metadata),
-        )
-        for dispatch in response.dispatches:
-            yield Dispatch.from_protobuf(dispatch)
+        while True:
+            response = await cast(
+                Awaitable[ListMicrogridDispatchesResponse],
+                self._stub.ListMicrogridDispatches(request, metadata=self._metadata),
+            )
+
+            yield [Dispatch.from_protobuf(dispatch) for dispatch in response.dispatches]
```

Comment:
You can potentially still do the `Dispatch` object construction lazily if you return a `Page` object. Or simply a generator:

```py
    async def list(...) -> AsyncIterator[Iterator[Dispatch]]:
        ...
        yield (Dispatch.from_protobuf(dispatch) for dispatch in response.dispatches)
```

With this if you do `first_dispatch = next(client.list(...))` only one `Dispatch` object will be ever created. Map will also work:

```py
yield map(Dispatch.from_protobuf, response.dispatches)
```

Reply:
I feel a bit more strong about this one, I think it makes more sense to return an iterator than a `list` here.